packages/amazon-bedrock/src/bedrock-chat-language-model.ts

```diff
@@ -0,0 +1,400 @@
+
+    const warnings: LanguageModelV1CallWarning[] = [];
+
+    if (frequencyPenalty != null) {
+      warnings.push({
+        type: 'unsupported-setting',
+        setting: 'frequencyPenalty',
+      });
+    }
+
+    if (presencePenalty != null) {
+      warnings.push({
+        type: 'unsupported-setting',
+        setting: 'presencePenalty',
+      });
+    }
+
+    if (seed != null) {
+      warnings.push({
+        type: 'unsupported-setting',
+        setting: 'seed',
+      });
+    }
+
+    const baseArgs: ConverseCommandInput = {
+      modelId: this.modelId,
+      system: this.settings.system,
+      additionalModelRequestFields: this.settings.additionalModelRequestFields,
+      additionalModelResponseFieldPaths:
+        this.settings.additionalModelResponseFields,
+      inferenceConfig: {
+        maxTokens: maxTokens,
+        temperature: temperature,
+        topP: topP,
+      },
+      messages: convertToBedrockChatMessages(prompt),
+    };
+
+    switch (type) {
+      case 'regular': {
+        return {
+          ...baseArgs,
+          ...prepareToolsAndToolChoice(mode),
+        } satisfies ConverseCommandInput;
+      }
+
+      case 'object-json': {
+        throw new UnsupportedFunctionalityError({
+          functionality: 'json-mode object generation',
+        });
+      }
+
+      case 'object-tool': {
+        return {
+          ...baseArgs,
+          toolConfig: {
+            tools: [
+              {
+                toolSpec: {
+                  name: mode.tool.name,
+                  description: mode.tool.description,
+                  inputSchema: { json: JSON.stringify(mode.tool.parameters) },
+                },
+              },
+            ],
+            toolChoice: { tool: { name: mode.tool.name } },
+          },
+        } satisfies ConverseCommandInput;
+      }
+
+      case 'object-grammar': {
+        throw new UnsupportedFunctionalityError({
+          functionality: 'grammar-mode object generation',
+        });
+      }
+
+      default: {
+        const _exhaustiveCheck: never = type;
+        throw new Error(`Unsupported type: ${_exhaustiveCheck}`);
+      }
+    }
+  }
+
+  async doGenerate(
+    options: Parameters<LanguageModelV1['doGenerate']>[0],
+  ): Promise<Awaited<ReturnType<LanguageModelV1['doGenerate']>>> {
+    const client = new BedrockRuntimeClient({
+      ...this.config,
+    });
+
+    const args = this.getArgs(options);
+
+    const response = await client.send(new ConverseCommand(args));
+
+    const { messages: rawPrompt, ...rawSettings } = args;
+
+    return {
+      text:
+        response.output?.message?.content
+          ?.map(part => part.text ?? '')
+          .join('') ?? undefined,
+      toolCalls: response.output?.message?.content
+        ?.filter(part => !!part.toolUse)
+        ?.map(part => ({
+          toolCallType: 'function',
+          toolCallId: part.toolUse?.toolUseId ?? generateId(),
+          toolName: part.toolUse?.name ?? 'unknown',
```

Comment:
when can this happen? can setting it to `unknown` lead to issues such as tools sharing the same name?

Reply:
strange, wonder when this could happen. how about `tool-${generateId()}`

Reply:
Updated